### DISCUSSION TITLE
Wiring diagram for control button / push button

### DISCUSSION THREAD

**Discussion by frankonski** [HomeSpan/HomeSpan#572] [Q&A] [answered]
Hey all. Everything was working fine with my little HomeSpan project. I set the status LED set with setStatusPin(5), all normal. I have another RGB LED for some other stuff, still all normal. Then I wired in a push button the same way as documented in the Freenove tutorial:
![image](https://user-images.githubusercontent.com/26729481/235355508-998d8080-cabd-4cb2-8ec2-6f72354ac6de.png)

At that point, I didn't have the homeSpan.setControlPin(13) code in yet, and as soon as I connected the board to its power source, I unfortunately fried it. Confirmed by Freenove Support. ESP32-WROVER dev board.

I ordered a new board, I'm back to a working state without the push button, and ready to re-wire it. There seems to be multiple ways to wire a push button, as per https://diyusthad.com/2019/05/3-different-way-of-connecting-push.html

What is the wiring expected for HomeSpan?

Thanks!

**Comment by HomeSpan** [comment]
You can connect a pushbutton between any ESP32 pin that is capable of being used as an input, and *either* ground *or* +3.3V (do not use 5V).  The SpanButton class allows you to specify whether you want to "trigger" on ground (LOW) or +3.3V (HIGH).  Note that the pushbutton should be of the type that is normally open (i.e. pressing the button completes the circuit).  As long as the ESP32 pin you are using has an internal pull-up/pull-down resistor (most do), then you do not need an external resistor.  But be careful not to select any ESP32 pins that might be output only (this depends on your specific board).

Based on the diagram above it seems you are using a 10-ohm resistor as a pull-up?  Is that correct, or am I reading the colors bands incorrectly?  If correct, that's way too low and is why the ESP32 fried.  You should be using 10K ohms or greater for external pull-ups.

**Reply by frankonski** [reply]
The resistors are 10K ohms. They came with the _starter kit_, and after I fried the board I tested them with a multimeter. They are good.
The pdf is at https://freenove.com/fnk0046/, page 46.
I tried the tutorial _Project 2.1 Button & LED_ some weeks back and it worked as expected.

I only want to use setControlPin to leverage the HomeSpan built-in features of reset-to-factory & all. Nothing fancy.

According to the doc, PIN 13 is good to use, and [this page](https://randomnerdtutorials.com/esp32-pinout-reference-gpios/) seems to imply that all but PINs 34, 35, 36 and 39 have internal pull-up/pull-down resistor.

But re-reading what you just said:

> You can connect a pushbutton between any ESP32 pin that is capable of being used as an input, and _**either**_ ground or +3.3V

.. if you look at the schema above, isn't the button connected to both ground **and** +3.3V?

Does HomeSpan attempt to automatically detect the controlPIN wiring, so to handle some internally-controlled _PushButton_ object and set pinMode(pinNumber, INPUT_PULLDOWN) as done [here](https://github.com/HomeSpan/HomeSpan/blob/master/src/Utils.cpp#L97) ?

**Comment by HomeSpan** [comment]
If you use SpanButton there is no need for any resistors.  HomeSpan automatically sets the pinMode to INPUT_PULLDOWN or INPUT_PULLUP depending on whether you specify a trigger on HIGH (pushbutton connects pin to +3.3V) or a trigger on LOW (pushbutton connects pin to ground).  Trigger on LOW (pushbutton connects pin to ground) is the default.

**Comment by HomeSpan** [comment, marked as answer]
Sorry, I missed the part where you said you were using the pushbutton as a homespan control pin.  You do not need to use SpanButton - HomeSpan *automatically* does this for you when you set a control pin.  Simply connect pushbutton to ground *without* any resistors.

**Reply by frankonski** [reply]
> Simply connect pushbutton to ground _without_ any resistors.

so this way?

`PIN13 -- pin(1) of pushbutton, pushbutton pin(2) -- ground`

? (as a said, really novice. Double-checking everything).

And out of curiosity, is it possible that my initial wiring of the pushButton, **without** calling homeSpan.setControlPin(13), may have fried my board? I won't sue you for 20$  ;-) .. you have been of great help so far!, but at least I will be able to continue with Freenove support, see if they have a way to unbrick it now that the cause would have been identified.

**Comment by HomeSpan** [comment]
Depending on how the pins are configured on start-up, it is possible that you were over-driving the current, but hard to know for sure.

Re your pushbutton pins, based on the diagram there are 4 pins, usually organized as two pairs.  Make sure you are not connecting the pins on the pushbutton that are not already connected on the same side of the button (if so, HomeSpan will think you are pressing and holding down the button).

You should be able to test which pins to use with a multimeter.

**Reply by frankonski** [reply]
Looks like that was it. Button is on and seems to be reacting the way it should.
I still don't understand what fried my previous board. I do power it from an external 12V source, with a 5V voltage regulator. But I am now starting to wonder if I mistakenly pluged in the 12V wire direct to the 3.3V breadboard line rather than the socket to the regulator... I can't go back in time, so I will never know. 
Anyhow, thanks for the quick help today!
